shared/js/dialer/contacts.js

```diff
@@ -11,6 +11,153 @@ var Contacts = {
     '/shared/js/fb/fb_reader_utils.js'
   ],
 
+  _calledNumberCache: [],
+
+  _CACHE_SIZE: 10,
+
+  init: function init() {
+    var self = this;
+    var mozContacts = window.navigator.mozContacts;
+
+    asyncStorage.getItem('contactsCache', function (value) {
+      if(value) {
+        self._calledNumberCache = JSON.parse(value);
+      }
+    });
+
+    if (mozContacts) {
+      LazyLoader.load(self._FB_FILES, function() {
+        mozContacts.oncontactchange = self._oncontactchange.bind(self);
+      });
+    }
+  },
+
+  _calledNumberCachePush: function (number, contact, matchingTel) {
+    if (this._calledNumberCache.length > this._CACHE_SIZE) {
+      this._calledNumberCache.shift();
+    }
+
+    this._calledNumberCache.push({
+      id: (contact ? contact.id : null),
+      number: number,
+      contact: contact,
+      matchingTel: matchingTel
+    });
+
+    var self = this;
+    setTimeout(function () {
+      asyncStorage.setItem('contactsCache', JSON.stringify(self._calledNumberCache));
+    },0);
+  },
+
+  _calledNumberCacheGet: function (number) {
```

Comment:
This and the ``init()`` function are not synchronized so it is possible that one might call this function before the contact cache has been populated. It's not a big deal because in that case we'd fall back on looking up the contact like we ordinarily do but it's worth keeping in mind.